Signal/src/environment/Environment.m

```diff
@@ -171,69 +172,76 @@ - (PropertyListPreferences *)preferences
+{
+    [self presentConversationForRecipientId:recipientId keyboardOnViewAppearing:compose callOnViewAppearing:NO];
 }
 
-+ (void)messageIdentifier:(NSString *)identifier withCompose:(BOOL)compose {
-    Environment *env          = [self getCurrent];
-    HomeViewController *vc = env.signalsViewController;
++ (void)callRecipientId:(NSString *)recipientId
+{
+    [self presentConversationForRecipientId:recipientId keyboardOnViewAppearing:NO callOnViewAppearing:YES];
+}
 
-    [[TSStorageManager sharedManager].dbReadWriteConnection
-        asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
-            TSThread *thread = [TSContactThread getOrCreateThreadWithContactId:identifier transaction:transaction];
-            [vc presentThread:thread keyboardOnViewAppearing:YES callOnViewAppearing:NO];
-        }];
++ (void)presentConversationForRecipientId:(NSString *)recipientId
+                  keyboardOnViewAppearing:(BOOL)keyboardOnViewAppearing
+                      callOnViewAppearing:(BOOL)callOnViewAppearing
+{
+    DispatchMainThreadSafe(^{
+        __block TSThread *thread = nil;
+        [[TSStorageManager sharedManager].dbReadWriteConnection
+            readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+                thread = [TSContactThread getOrCreateThreadWithContactId:recipientId transaction:transaction];
+            }];
+        [self presentConversationForThread:thread
+                   keyboardOnViewAppearing:keyboardOnViewAppearing
+                       callOnViewAppearing:callOnViewAppearing];
+    });
 }
 
-+ (void)callUserWithIdentifier:(NSString *)identifier
++ (void)presentConversationForThread:(TSThread *)thread
 {
-    Environment *env = [self getCurrent];
-    HomeViewController *vc = env.signalsViewController;
-
-    [[TSStorageManager sharedManager].dbReadWriteConnection
-        asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
-            TSThread *thread = [TSContactThread getOrCreateThreadWithContactId:identifier transaction:transaction];
-            [vc presentThread:thread keyboardOnViewAppearing:NO callOnViewAppearing:YES];
-        }];
+    [self presentConversationForThread:thread keyboardOnViewAppearing:YES callOnViewAppearing:NO];
 }
 
-+ (void)messageGroup:(TSGroupThread *)groupThread {
-    Environment *env          = [self getCurrent];
-    HomeViewController *vc = env.signalsViewController;
++ (void)presentConversationForThread:(TSThread *)thread
+             keyboardOnViewAppearing:(BOOL)keyboardOnViewAppearing
+                 callOnViewAppearing:(BOOL)callOnViewAppearing
+{
+    if (!thread) {
+        OWSFail(@"%@ Can't present nil thread.", self.tag);
+        return;
+    }
+
+    DispatchMainThreadSafe(^{
+        UIViewController *frontmostVC = [[UIApplication sharedApplication] frontmostViewController];
+
+        if ([frontmostVC isKindOfClass:[ConversationViewController class]]) {
```

Comment:
This "don't navigate to thread if we're already showing the thread" logic now applies to contact and groups threads and applies to all scenarios in which we try to navigate to a conversation, including placing calls, etc.